### ISSUE TITLE
`SendEmail` error: `HTTP/1.1 505 HTTP Versi﻿on not supported` error

### ISSUE THREAD

**Issue by edwinyzh** [gmurt/ksAws#2]
When calling the Indy-powered `SendEmail` I'm getting the `HTTP/1.1 505 HTTP Versi﻿on not supported` error.

then I added the following line to `TksAwsBaseService.GetUrl`:
`  Result := UrlEncode(Result);`
And now I'm getting the following new error:
`HTTP/1.1 403 Forbidden.`

Any idea? Thanks!

**Comment by edwinyzh** [comment]
@gmurt,

It's strange, I'm using the same IAM user, it was ok when I tried several days ago with the 'VerifiedSenders' sample project. But now I tried again with the same set of access key and secret, I'm getting the `HTTP/1.1 403 Forbidden` error as well...

In [this thread](https://forums.aws.amazon.com/thread.jspa?messageID=233422&#233422) it seems that AWS returns details about the error.  With ksAws is there a way to check detailed error like that?

Thanks!

**Comment by edwinyzh** [comment]
Update 1: I confirm with the AWS CLI, the `aws ses list-identities` command successfully returned the verified email addresses. In other words, my IAM user access id and key are correct.

**Comment by gmurt** [comment]
Just pushed an update, it looks like it may have been to do with query/header encoding.  Let me know if it's fixed, appears to be working for me now.

**Comment by edwinyzh** [comment]
Thanks @gmurt 

To get the latest commit compiled, I had to change in ksAwsHash.pas the following line
`  C_UNSAFE_CHARS: array of Char = [' ',':', '/', '?', '#', '[', ']', '@', '!', '$', '&', '''', '(', ')', '*', '+', ',', ';', '='];`
to
`  C_UNSAFE_CHARS: array[0..18] of Char = (' ',':', '/', '?', '#', '[', ']', '@', '!', '$', '&', '''', '(', ')', '*', '+', ',', ';', '=');`

And unfortunately, I'm still getting the `HTTP/1.1 403 Forbidden.` error with both `SendEmail` and `GetSenders`.

I attached the testing code below:
[TestKsAwsSesEmailSending_tmp.zip](https://github.com/gmurt/ksAws/files/5459842/TestKsAwsSesEmailSending_tmp.zip)

**Comment by gmurt** [comment]
Both demo functions are working fine for me in XE4. 403 does indicate a credential issue, maybe declare your credentials as constants to ensure they do not have any spaces at the beginning or end. 

![image](https://user-images.githubusercontent.com/1161351/97609840-045c0000-1a0c-11eb-983e-152037fa6e2e.jpeg)

**Comment by edwinyzh** [comment]
Thanks @gmurt ! Good to know it works, I'll try again.
Another think I want to confirm - were you using **Indy** as the HTTP client to test?

**Comment by edwinyzh** [comment]
@gmurt,

With the same testing project I attached without further changes, after restarted my computer, now the btnGeVerifiedSenders can retrieve the correct list of senders.

But the btnSend still failed, with **a different error**: `HTTP/1.1 400 Bad Request.`

**Comment by edwinyzh** [comment]
Hey Graham,

With the same testing project I attached without further changes, after **restarted** my computer, now both the btnGeVerifiedSenders and btnSend are working!!!

Not sure why, maybe the constant switching of "direct connection vs shadowsocks (proxy) connection" on the computer caused the issue...

**Comment by gmurt** [comment]
Good to hear. Any further problems let me know.

**Comment by edwinyzh** [comment]
@gmurt,

I'm getting the `HTTP/1.1 505 HTTP Version not supported` error again, maybe due to my long email body text - over 1600 characters (260 words)? Because I guess the email message gets sent through URL? That'll exceed the length limit I think.

I think this kind of email body length is very normal.

**Comment by edwinyzh** [comment]
Hi Graham,
After further testing, I found that the **content** of the email subject and body affects if the error will happen - email can be sent with simple subject and email body like the code I attached, but I'll get 403 or 505 error depending on the content of subject and/or body.

For example, the following code will work:
```
    msg := CreateSesMessage('edwin@a.com', 'john@b.com',
      'testing AWS SES email sending - '  + DateTimeToStr(Now),
      'the email was sent at: ' + DateTimeToStr(Now));
```
The following code does NOT work (notice the **double quotes** in the subject), it will produce the `EIdHTTPProtocolException: HTTP/1.1 403 Forbidden` error:
```
    msg := CreateSesMessage('edwin@a.com', 'john@b.com',
      'testing "AWS SES" email sending - '  + DateTimeToStr(Now),
      'the email was sent at: ' + DateTimeToStr(Now));

```

**Comment by gmurt** [comment]
Ok, I'll take a look. Can you also provide me a simple example which produces the 505 error on your machine?

This will help me identify the issue.

**Comment by gmurt** [comment]
I believe the latest update will fix the 403 error when using Indy. If you can provide an example which produces the 505 that would help me fix that one.

**Comment by edwinyzh** [comment]
I confirm with the latest update I'm no longer getting the 403 error, but only the `EIdHTTPProtocolException: HTTP/1.1 505 HTTP Version not supported` error. I will provide an example.

**Comment by edwinyzh** [comment]
@gmurt , the following code will produce the 505 error. The code example was extracted from the sample I attached earlier - I just added double quotes to the email subject.

BTW, WWIW, I'm using a latest revision of Indy from the official repository: https://github.com/IndySockets/Indy

```
procedure TfrmMain.btnSendClick(Sender: TObject);
var
  msg: IksAwsSesMessage;
  ses: IksAwsSES;
begin
  ses := CreateSes(edtAccessKeyId.Text, edtAccessSecret.Text, awsUsEast1);

  // ok:
  //  msg := CreateSesMessage('edwin.yip@mydomain.com', 'anotheruser@mydomain.com',
  //    'testing AWS SES email sending - '  + DateTimeToStr(Now),
  //    'the email was sent at: ' + DateTimeToStr(Now));

  // will produce the  "EIdHTTPProtocolException: HTTP/1.1 505 HTTP Version not supported" error:
  msg := CreateSesMessage('edwin.yip@mydomain.com', 'anotheruser@mydomain.com',
    'testing "AWS SES" email sending - '  + DateTimeToStr(Now),
    'the email was sent at: ' + DateTimeToStr(Now));

  ses.SendEmail(msg);
end;

```

**Comment by gmurt** [comment]
Think it was the encoding of the space " " character.  Please pull latest changes and retry.

**Comment by edwinyzh** [comment]
> Think it was the encoding of the space " " character. Please pull latest changes and retry.

Thanks Graham! I'll check it now. BTW, does `SendEmail` support html message? Thanks!

**Comment by gmurt** [comment]
Not sure, I've only tested with plain text messages. I'll close this issue, if there's a problem with html please open a new ticket and I'll take a look.

**Comment by gmurt** [comment]
Sorry I mis-read your last message. Thought you confirmed it was fixed. Please test and if the 505 is fixed I'll close this issue.

**Comment by edwinyzh** [comment]
@gmurt,

The issue with double quotes in the subject has been fixed, but with body text like the following still emit the `EIdHTTPProtocolException: HTTP/1.1 505 HTTP Version not supported` error:

```
  cTestBody =
    'Dear E,' + sLineBreak +
    '' + sLineBreak +
    'This the “re-bbb b”.' + sLineBreak +
    'I''m - didn''t doesn’t alter.' + sLineBreak +
    '' + sLineBreak +
    'policy:' + sLineBreak +
    '-	Wrs.' + sLineBreak +
    '-	Wres.' + sLineBreak +
    '-	Dt 2 mo(plo).' + sLineBreak +
    '-	If you on: UUUIAUSFD (val, 30 cent dy-di ‘test’. on). ' + sLineBreak +
    '' + sLineBreak +
    'Here:' + sLineBreak +
    '-	download from https://google.com' + sLineBreak +
    '-	key: john-jdo#001-01201030-c826-e061-9208-50ad-7c56-6629-023c-bb5c' + sLineBreak +
    '' + sLineBreak +
    'I .. didn’t hap questions.' + sLineBreak +
    '' + sLineBreak +
    'Edwin Doe – the of aa/bb' + sLineBreak +
    'Best Regards.';
var
  msg: IksAwsSesMessage;
begin
    msg := CreateSesMessage(cToEmail, cSendFrom, cSubject, cTestBody);
    ses.SendEmail(msg);
end;
```

**Comment by gmurt** [comment]
Please can you supply me the full source for the unit so I can see how the constants are declared.

**Comment by edwinyzh** [comment]
Attached.
[TestKsAwsSesEmailSending002.zip](https://github.com/gmurt/ksAws/files/5469021/TestKsAwsSesEmailSending002.zip)

**Comment by gmurt** [comment]
Ok, hopefully now sorted! Let me know if it's fixed with the latest change.

**Comment by edwinyzh** [comment]
Thanks again for the help @gmurt.
Unfortunately, with the same example without any code changes nor other changes on AWS, but just a rebuild, Now I get:
`HTTP/1.1 403 Forbidden.`

**Comment by gmurt** [comment]
Working fine here. Forbidden suggests it may be an issue with your credentials.  Ensure that you are entering them without spaces etc.  Just tested again with XE4/Indy. I'll try and make some improvements so more detailed error messages are displayed.

**Comment by gmurt** [comment]
Just googled, a 403 suggests that there may be a leading/trailing space added by mistake to your secret key. I tested and if I add a space to the Secret key I do get the 403 error.  I'll make sure the keys are "trimmed" with the next update but please check this.

**Comment by edwinyzh** [comment]
Hello Graham,

I confirm there are no leading/trailing spaces added to the keys. It's the same test program we've been using in the past days. the 'subject' and 'body' were the only things that were changed during the test.

**Comment by gmurt** [comment]
I've just resynced the repo to my machine and run your example in XE4 with the standard Indy which ships with that version. 

The only change I've made is to hard code the credentials and changed the email addresses. It works fine, it "has" to be a credential issue at your end, there's nothing else it can be. Maybe create some new credentials in IAM and test in case your existing credentials have been invalidated.

![image](https://user-images.githubusercontent.com/1161351/97786822-1671a600-1ba6-11eb-82d5-e497f37063fe.jpeg)


![image](https://user-images.githubusercontent.com/1161351/97786848-41f49080-1ba6-11eb-8f14-663a810fe76d.jpeg)

**Comment by edwinyzh** [comment]
@gmurt ,

Thanks for your help and sorry for all the troubles - not sure why, but today I re-run the programs (both, one real project, another is the test program I sent to you), without any changes, I tried several times, and the email sending **worked**!

I'm also puzzled. The only thing that's changed is the computer has been rebooted. But last night (my time) I did try to reboot the PC and try again, but the problem existed.
